_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+The CHERIoT ISA is tailored for a specific compartmentalisation model and core software stack, which is how we're able to provide usable abstractions for compartmentalisation on such tiny hardware.
+A lot of other work on CHERI has focused more on big systems and incremental migration of truly huge software stacks such as the Chromium web browser running on a fully memory-safe kernel, display server, and so on.
+
+Back in 2015, we thought 15 coauthors was a lot!
+The [preceding 2014 paper](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201503-asplos2015-cheri-cmachine.pdf) had only nine coauthors, so I think this was the one where we discovered that the ACM LaTeX style had a hard-coded limit of 12 authors (and if you had that many, you ended up with an entire page listing the authors).
+The [most recent CHERI technical report](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-987.pdf) has 27 coauthors.
+
+# A lot is still the same
+
+Looking back at the 2015 paper, it's somewhat surprising how recognisable the system it describes would be for someone who is familiar with CHERIoT.
+A huge number of the features that make CHERIoT possible were there already.
+In particular, we'd already made two of the biggest changes from the first version of CHERI that I had used.
+
+In the very first CHERI version, capability bounds were represented by a base and a top, but there was no separate address.
+In C, pointer arithmetic would move the base (truncating the capability) so you often needed to carry a separate offset.
+Around 2014, we changed this so the capability stored an offset instead.
+Microarchitecturally, this was actually an address, but we exposed it in the ISA as an offset from the base.
+
+Exposing this address as an offset was my idea.
+I hoped that we would be able to create a C dialect where copying garbage collectors were possible.
+Any conversion from a pointer to an integer gave the offset unless you intentionally asked for the address (in which case it was your responsibility to ensure that you didn't rely on that address being stable across GC sweeps).
+This meant that (in theory) you could relocate any object and update all capabilities to it (and, remember, the tag bit means that you *can* find all capabilities).
+I had a student, Munraj Vadera, write a copying GC for C and it worked!
+It could even do sub-object collection, so if you allocated an array and kept a bounded capability to a single element, the rest of the array could be collected.
+
+I was very happy with this model.
+Unfortunately, it turned out to be a terrible idea.
+When Alex Richardson and Jessica Clarke tried to build much more complex programs with it, they discovered that lots of things really want to use pointer to integer conversions for things like trees and hash tables.
+There were some really nasty corner cases, such as hash tables that 'worked', but ended up putting every object in the first hash bucket, giving truly awful performance.
```

Comment:
More complex problems such as lock ordering based on pointer address to the object are also sad.